### PR TITLE
8338569: HTTP/1.1 CleanupTrigger may be triggerred after the next exchange started

### PR DESCRIPTION
**Pull request by dfuch**
The java/net/httpclient/DigestEchoClient.java fails intemittently with IOException: HTTP/1.1 header parser received no bytes.

Analysis shows that this is caused by the CleanupTrigger which receives data after the reused connection has been taken out of the HTTP/1.1 clear pool. This should not happen.

The issue is caused by deferred registration of read subscribers with the SocketTube. The subscribers are registered within the SelectorManager thread to ensure proper interaction with the read method.

Sometimes pending subscribers are pushed faster than they are actually subscribed, which may cause the wrong subscriber to be subscribed at the wrong time. 

This is a redo of [JDK-8336655](https://bugs.openjdk.org/browse/JDK-8336655) which was a failed fix.
The main difference is that the new fix uses a ConcurrentLinkedQueue to make sure all subscribers get subscribed and dropped in the right sequence. Marking previous subscription as stopped ensures that data can only be routed to the last subscriber in the queue.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8338569](https://bugs.openjdk.org/browse/JDK-8338569): HTTP/1.1 CleanupTrigger may be triggerred after the next exchange started (**Bug** - P4)

### Reviewers
 * [Jaikiran Pai](https://openjdk.org/census#jpai) (@jaikiran - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/20623/head:pull/20623` \
`$ git checkout pull/20623`

Update a local copy of the PR: \
`$ git checkout pull/20623` \
`$ git pull https://git.openjdk.org/jdk.git pull/20623/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 20623`

View PR using the GUI difftool: \
`$ git pr show -t 20623`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/20623.diff">https://git.openjdk.org/jdk/pull/20623.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/20623#issuecomment-2296135628)

### CODE REVIEW

**Comment by dfuch** [comment]
Note: the changes to ShutdownNow.java are coincidental. I observed this test failing once in my many tests, and decided to make the exception it throws more explicit. FWIW - I had also observed this test failing before the change, it's intermittent and rare. I took the opportunity to raise the timeout from 2000 to 2500.

File: src/java.net.http/share/classes/jdk/internal/net/http/ConnectionPool.java [hunk 1]
```diff
@@ -524,18 +525,37 @@ private boolean contains0(HttpConnection c) {
...
+            if (Log.errors()) {
+                Log.logError("WARNING: CleanupTrigger triggered for" +
+                        " a connection not found in the pool: closing {0}", c.dbgString());
+            }
+            if (debug.on()) {
+                debug.log("WARNING: CleanupTrigger triggered for" +
+                        " a connection not found in the pool: closing %s", c.dbgString());
+            }
+            Throwable cause = new IOException("Unexpected cleanup triggered for non pooled connection", error);
+            cause.printStackTrace();
```
Comment (jaikiran) [inline src/java.net.http/share/classes/jdk/internal/net/http/ConnectionPool.java:554]: Is the stacktrace printing intentional?

**Review by jaikiran — COMMENTED** [review]
This looks OK to me. Some of the files will need a copyright year update before integrating.

File: src/java.net.http/share/classes/jdk/internal/net/http/ConnectionPool.java (hunk 1 above)
Comment (dfuch) [reply src/java.net.http/share/classes/jdk/internal/net/http/ConnectionPool.java:554]: Oh darn! That's a left over. Thanks for catching it.

**Comment by dfuch** [comment]
Copyright years have been updated and the stray printStackTrace() has been removed.

**Review by jaikiran — APPROVED** [review]

**Comment by dfuch** [comment]
Thanks Jaikiran. I will integrate tomorrow.

**Pull request closed without merging**